clippy_lints/src/doc/broken_link.rs

```diff
@@ -0,0 +1,112 @@
+// - When a link tag is missing the close parenthesis character at the end.
+// - When a link has whitespace within the url link.
+enum BrokenLinkReason {
+    MultipleLines,
+}
+
+#[derive(Debug)]
+enum State {
+    ProcessingLinkText,
+    ProcessedLinkText,
+    ProcessingLinkUrl(UrlState),
+}
+
+#[derive(Debug)]
+enum UrlState {
+    Empty,
+    FilledEntireSingleLine,
+    FilledBrokenMultipleLines,
+}
+
+fn warn_if_broken_link(cx: &LateContext<'_>, bl: &PullDownBrokenLink<'_>, doc: &String, fragments: &Vec<DocFragment>) {
+    if let Some(span) = source_span_for_markdown_range(cx.tcx, doc, &bl.span, fragments) {
+        // `PullDownBrokenLink` provided by pulldown_cmark always
+        // start with `[` which makes pulldown_cmark consider this a link tag.
+        let mut state = State::ProcessingLinkText;
+
+        // Whether it was detected a line break within the link tag url part.
+        let mut reading_link_url_new_line = false;
+
+        // Skip the first char because we already know it is a `[` char.
+        for (abs_pos, c) in doc.char_indices().skip(bl.span.start + 1) {
+            match &state {
+                State::ProcessingLinkText => {
+                    if c == ']' {
+                        state = State::ProcessedLinkText;
+                    }
+                },
+                State::ProcessedLinkText => {
+                    if c == '(' {
+                        state = State::ProcessingLinkUrl(UrlState::Empty);
+                    } else {
+                        // not a real link, just skip it without reporting a broken link for it.
+                        break;
+                    }
+                },
+                State::ProcessingLinkUrl(url_state) => {
+                    if c == '\n' {
+                        reading_link_url_new_line = true;
+                        continue;
+                    }
+
+                    if c == ')' {
+                        // record full broken link tag
+                        if let UrlState::FilledBrokenMultipleLines = url_state {
+                            let offset = abs_pos - bl.span.start;
+                            report_broken_link(cx, span, offset, BrokenLinkReason::MultipleLines);
+                        }
+                        break;
+                    }
+
+                    if !c.is_whitespace() {
+                        if reading_link_url_new_line {
+                            // It was reading a link url which was entirely in a single line, but a new char
+                            // was found in this new line which turned the url into a broken state.
+                            state = State::ProcessingLinkUrl(UrlState::FilledBrokenMultipleLines);
+                            continue;
+                        }
+
+                        state = State::ProcessingLinkUrl(UrlState::FilledEntireSingleLine);
+                    }
+                },
+            };
+        }
+    }
+}
```

Comment:
This looks like this is basically:
* `strip_prefix('[')`
* `split_once(']')`
* `strip_prefix('(')`
* `find(['\n', ')'])` until the end is reached.

It would be nice if it did that instead of using an explicit state machine.

---

Some other things:
* Is there a reason to not also lint if there's a line break between the link text in the URL?
* This could probably be brought to zero false positives if the url part is actually checked for being a valid url.

Reply:
Good suggestion about dropping the state machine approach. It does looks simpler now.

About your question:

> Is there a reason to not also lint if there's a line break between the link text in the URL?

`rustdoc` doesn't break in that case, so the checker doesn't warn about it. It has a test for that already, checkout the test for `doc_valid_link_broken_text`.

About this other comment:

> This could probably be brought to zero false positives if the url part is actually checked for being a valid url.

I added a very simple http step verification, which should be enough I think.